File: skills/obercreate/scripts/aggregate_benchmark.py
#!/usr/bin/env python3
"""Aggregate benchmark results from graded eval runs.

Reads grading.json and timing.json files from a benchmark directory structure,
computes summary statistics per configuration, and outputs benchmark.json and
benchmark.md.

Directory structure expected:
    benchmark_dir/
        eval-name/
            eval_metadata.json  (optional)
            with_skill/
                run-1/grading.json, timing.json
                run-2/grading.json, timing.json
            without_skill/
                run-1/grading.json, timing.json
"""

import argparse
import json
import math
import sys
from datetime import datetime, timezone
from pathlib import Path


def calculate_stats(values):
    """Compute mean, sample stddev, min, max for a list of numbers.

    Uses sample standard deviation (n-1 denominator) when len > 1.
    Returns zeros for empty input.
    """
    if not values:
        return {"mean": 0, "stddev": 0, "min": 0, "max": 0}

    mean = sum(values) / len(values)

    if len(values) > 1:
        variance = sum((x - mean) ** 2 for x in values) / (len(values) - 1)
        stddev = math.sqrt(variance)
    else:
        stddev = 0

    return {
        "mean": round(mean, 4),
        "stddev": round(stddev, 4),
        "min": round(min(values), 4),
        "max": round(max(values), 4),
    }


def load_run_results(benchmark_dir):
    """Walk benchmark directory and collect grading results from each run.

    Returns a list of run dicts with eval metadata, configuration, results,
    expectations, and user notes.
    """
    runs = []

    for eval_dir in sorted(benchmark_dir.iterdir()):
        if not eval_dir.is_dir():
            continue
        if eval_dir.name.startswith("."):
            continue

        # Read optional eval metadata
        metadata_path = eval_dir / "eval_metadata.json"
        eval_id = None
        eval_name = eval_dir.name

        if metadata_path.exists():
            meta = json.loads(metadata_path.read_text())
            eval_id = meta.get("eval_id", eval_name)
            eval_name = meta.get("eval_name", eval_name)

        # Each subdirectory is a configuration (with_skill, without_skill, etc.)
        for config_dir in sorted(eval_dir.iterdir()):
            if not config_dir.is_dir():
                continue

            configuration = config_dir.name

            # Look for run-N subdirectories; fall back to config_dir as single run
            run_dirs = sorted(
                d for d in config_dir.iterdir()
                if d.is_dir() and d.name.startswith("run")
            )
            if not run_dirs:
                run_dirs = [config_dir]

            for run_num, run_dir in enumerate(run_dirs, 1):
                grading_path = run_dir / "grading.json"
                if not grading_path.exists():
                    grading_path = config_dir / "grading.json"
                if not grading_path.exists():
                    print(
                        f"Warning: no grading.json in {run_dir}",
                        file=sys.stderr,
                    )
                    continue

                grading = json.loads(grading_path.read_text())
                summary = grading.get("summary", {})

                # Timing: check run dir first, then config dir
                timing_path = run_dir / "timing.json"
                if not timing_path.exists():
                    timing_path = config_dir / "timing.json"
                timing = (
                    json.loads(timing_path.read_text())
                    if timing_path.exists()
                    else {}
                )

                metrics = grading.get("execution_metrics", {})

                run = {
                    "eval_id": eval_id or eval_name,
                    "eval_name": eval_name,
                    "configuration": configuration,
                    "run_number": run_num,
                    "result": {
                        "pass_rate": summary.get("pass_rate", 0),
                        "passed": summary.get("passed", 0),
                        "total": summary.get("total", 0),
                        "time_seconds": timing.get(
                            "total_duration_seconds", 0
                        ),
                        "tokens": timing.get("total_tokens", 0),
                        "tool_calls": metrics.get("total_tool_calls", 0),
                        "errors": metrics.get("errors_encountered", 0),
                    },
                    "expectations": grading.get("expectations", []),
                    "notes": [],
                }

                notes_summary = grading.get("user_notes_summary", {})
                if notes_summary:
                    run["notes"] = notes_summary.get(
                        "uncertainties", []
                    ) + notes_summary.get("workarounds", [])

                runs.append(run)

    return runs


def aggregate_results(runs):
    """Compute per-configuration summary statistics and delta between first two configs."""
    by_config = {}
    for run in runs:
        config = run["configuration"]
        if config not in by_config:
            by_config[config] = {"pass_rates": [], "times": [], "tokens": []}
        by_config[config]["pass_rates"].append(run["result"]["pass_rate"])
        by_config[config]["times"].append(run["result"]["time_seconds"])
        by_config[config]["tokens"].append(run["result"]["tokens"])

    summary = {}
    for config, data in by_config.items():
        summary[config] = {
            "pass_rate": calculate_stats(data["pass_rates"]),
            "time_seconds": calculate_stats(data["times"]),
            "tokens": calculate_stats(data["tokens"]),
        }

    # Delta between first two configurations
    configs = list(by_config.keys())
    if len(configs) >= 2:
        a, b = configs[0], configs[1]
        summary["delta"] = {
            "pass_rate": (
                f"{summary[a]['pass_rate']['mean'] - summary[b]['pass_rate']['mean']:+.4f}"
            ),
            "time_seconds": (
                f"{summary[a]['time_seconds']['mean'] - summary[b]['time_seconds']['mean']:+.1f}"
            ),
            "tokens": (
                f"{summary[a]['tokens']['mean'] - summary[b]['tokens']['mean']:+.0f}"
            ),
        }

    return summary


def generate_markdown(benchmark):
    """Render benchmark data as a Markdown report."""
    lines = [
        f"# Benchmark: {benchmark['metadata']['skill_name']}",
        "",
        f"**Generated:** {benchmark['metadata']['timestamp']}",
        "",
    ]

    summary = benchmark.get("run_summary", {})

    for config, stats in summary.items():
        if config == "delta":
            continue
        lines.append(f"## {config}")
        lines.append("| Metric | Mean | Stddev | Min | Max |")
        lines.append("|--------|------|--------|-----|-----|")
        for metric in ["pass_rate", "time_seconds", "tokens"]:
            s = stats.get(metric, {})
            lines.append(
                f"| {metric} | {s.get('mean', 0)} | {s.get('stddev', 0)} "
                f"| {s.get('min', 0)} | {s.get('max', 0)} |"
            )
        lines.append("")

    if "delta" in summary:
        lines.append("## Delta")
        for k, v in summary["delta"].items():
            lines.append(f"- {k}: {v}")
        lines.append("")

    if benchmark.get("notes"):
        lines.append("## Notes")
        for note in benchmark["notes"]:
            lines.append(f"- {note}")
        lines.append("")

    return "\n".join(lines)


def main():
    parser = argparse.ArgumentParser(
        description="Aggregate benchmark results from graded eval runs"
    )
    parser.add_argument(
        "benchmark_dir",
        help="Directory containing eval result subdirectories",
    )
    parser.add_argument(
        "--skill-name",
        required=True,
        help="Name of the skill being benchmarked",
    )
    parser.add_argument(
        "--skill-path",
        default="",
        help="Path to the skill's SKILL.md (optional)",
    )
    parser.add_argument(
        "--output",
        help="Output directory (default: benchmark_dir)",
    )
    args = parser.parse_args()

    benchmark_dir = Path(args.benchmark_dir)
    output_dir = Path(args.output) if args.output else benchmark_dir

    runs = load_run_results(benchmark_dir)
    if not runs:
        print("Error: no grading results found", file=sys.stderr)
        sys.exit(1)

    run_summary = aggregate_results(runs)

    configurations = set(r["configuration"] for r in runs)
    runs_per_config = max(
        (len([r for r in runs if r["configuration"] == c]) for c in configurations),
        default=0,
    )

    benchmark = {
        "metadata": {
            "skill_name": args.skill_name,
            "skill_path": args.skill_path,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "evals_run": sorted(set(r["eval_name"] for r in runs)),
            "runs_per_configuration": runs_per_config,
        },
        "runs": runs,
        "run_summary": run_summary,
        "notes": [],
    }

    output_dir.mkdir(parents=True, exist_ok=True)
    (output_dir / "benchmark.json").write_text(
        json.dumps(benchmark, indent=2)
    )
    (output_dir / "benchmark.md").write_text(generate_markdown(benchmark))

    print(f"Wrote benchmark.json and benchmark.md to {output_dir}")


if __name__ == "__main__":
    main()

File: mcp/src/tools/aggregate-benchmark.ts
/**
 * aggregate_benchmark — pure TS aggregation of graded runs into benchmark.json
 * + benchmark.md. The delta is numeric and between explicitly NAMED configs
 * (candidate − baseline); shipping gates are computed in code, never self-assessed.
 * Returns notes: string[] only — interpretive pattern analysis belongs to a
 * fresh-context analyzer reading benchmark.json, not to this tool.
 */
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { join, resolve } from "node:path";
import { z } from "zod";
import { calculateStats, delta } from "../lib/stats.ts";
import { ok, err, type ToolModule, type ToolResult } from "../lib/tool.ts";
import { walkIterationDir } from "../lib/workspace.ts";
import type { Benchmark, BenchmarkRun, Expectation } from "../types.ts";

export const name = "aggregate_benchmark";
export const title = "Aggregate graded runs into benchmark.json + benchmark.md";
export const description =
  "Walks an iteration directory of graded runs, computes mean / sample-stddev (n-1) / min / max for " +
  "pass_rate, time_seconds, tokens, and cost_usd per configuration, and a numeric delta between two " +
  "explicitly named configurations (candidate minus baseline — never whichever two configs happened to " +
  "come first). Also evaluates the shipping gates in code: pressure_adherence (every pressure-eval " +
  "candidate-config run COMPLIANT — the '100% workflow adherence under 3+ pressure factors' criterion) " +
  "and skill_lift (candidate pass_rate mean above baseline). Writes benchmark.json and benchmark.md into " +
  "the iteration dir. Run after run_eval has covered every eval; dispatch a fresh-context analyzer over " +
  "benchmark.json for pattern findings.";

export const inputShape = {
  iteration_dir: z.string().describe("Path to <skill>-workspace/iteration-N/."),
  skill_name: z.string(),
  skill_path: z.string().optional(),
  baseline_config: z.string().default("without_skill"),
  candidate_config: z.string().default("with_skill"),
  response_format: z.enum(["concise", "detailed"]).default("concise"),
};

type Input = z.output<z.ZodObject<typeof inputShape>>;

function num(o: Record<string, unknown> | undefined, k: string): number {
  return typeof o?.[k] === "number" ? (o[k] as number) : 0;
}

function generateMarkdown(b: Benchmark): string {
  const lines = [
    `# Benchmark: ${b.metadata.skill_name}`,
    "",
    `**Generated:** ${b.metadata.timestamp}`,
    `**Evals:** ${b.metadata.evals_run.join(", ")}`,
    "",
  ];
  for (const [config, stats] of Object.entries(b.run_summary)) {
    lines.push(`## ${config}`, "| Metric | Mean | Stddev | Min | Max |", "|--------|------|--------|-----|-----|");
    for (const metric of ["pass_rate", "time_seconds", "tokens", "cost_usd"] as const) {
      const s = stats[metric];
      lines.push(`| ${metric} | ${s.mean} | ${s.stddev} | ${s.min} | ${s.max} |`);
    }
    lines.push("");
  }
  if (b.delta) {
    lines.push(`## Delta (${b.delta.candidate} − ${b.delta.baseline})`);
    const sign = (n: number): string => (n >= 0 ? `+${n}` : `${n}`);
    lines.push(`- pass_rate: ${sign(b.delta.pass_rate)}`);
    lines.push(`- time_seconds: ${sign(b.delta.time_seconds)}`);
    lines.push(`- tokens: ${sign(b.delta.tokens)}`);
    lines.push(`- cost_usd: ${sign(b.delta.cost_usd)}`);
    lines.push("");
  }
  lines.push("## Gates");
  lines.push(`- pressure_adherence: ${b.gates.pressure_adherence === null ? "n/a (no pressure evals)" : b.gates.pressure_adherence}`);
  lines.push(`- skill_lift: ${b.gates.skill_lift === null ? "n/a" : b.gates.skill_lift}`);
  lines.push("");
  if (b.notes.length > 0) {
    lines.push("## Notes");
    for (const note of b.notes) lines.push(`- ${note}`);
    lines.push("");
  }
  return lines.join("\n");
}

export async function handler(args: Input): Promise<ToolResult> {
  const iterDir = resolve(args.iteration_dir);
  if (!existsSync(iterDir)) return err(`iteration_dir does not exist: ${iterDir}`);

  const { runs: walked, notes } = walkIterationDir(iterDir);
  if (walked.length === 0) return err(`no graded runs found under ${iterDir}`);

  const runs: BenchmarkRun[] = walked.map((w) => {
    const summary = (w.grading.summary ?? {}) as Record<string, unknown>;
    const metrics = (w.grading.execution_metrics ?? {}) as Record<string, unknown>;
    const pressure = w.grading.pressure_compliance as Record<string, unknown> | undefined;
    const notesSummary = (w.grading.user_notes_summary ?? {}) as Record<string, unknown>;
    const runNotes = [
      ...(Array.isArray(notesSummary.uncertainties) ? (notesSummary.uncertainties as string[]) : []),
      ...(Array.isArray(notesSummary.workarounds) ? (notesSummary.workarounds as string[]) : []),
    ];
    const metricsFile = join(w.run_dir, "metrics.json");
    let costUsd = 0;
    if (existsSync(metricsFile)) {
      try {
        const m: unknown = JSON.parse(readFileSync(metricsFile, "utf8"));
        if (typeof m === "object" && m !== null) costUsd = num(m as Record<string, unknown>, "total_cost_usd");
      } catch {
        // ignore unreadable metrics
      }
    }
    return {
      eval_id: w.eval_id,
      eval_name: w.eval_name,
      configuration: w.configuration,
      run_number: w.run_number,
      result: {
        pass_rate: num(summary, "pass_rate"),
        passed: num(summary, "passed"),
        total: num(summary, "total"),
        time_seconds: w.timing.total_duration_seconds,
        tokens: w.timing.total_tokens,
        cost_usd: costUsd,
        tool_calls: num(metrics, "total_tool_calls"),
        errors: num(metrics, "errors_encountered"),
      },
      expectations: Array.isArray(w.grading.expectations) ? (w.grading.expectations as Expectation[]) : [],
      pressure_verdict: typeof pressure?.verdict === "string" ? (pressure.verdict as string) : null,
      notes: runNotes,
    };
  });

  // Per-config summary stats.
  const byConfig = new Map<string, BenchmarkRun[]>();
  for (const r of runs) {
    const list = byConfig.get(r.configuration) ?? [];
    list.push(r);
    byConfig.set(r.configuration, list);
  }
  const runSummary: Benchmark["run_summary"] = {};
  for (const [config, list] of byConfig) {
    runSummary[config] = {
      pass_rate: calculateStats(list.map((r) => r.result.pass_rate)),
      time_seconds: calculateStats(list.map((r) => r.result.time_seconds)),
      tokens: calculateStats(list.map((r) => r.result.tokens)),
      cost_usd: calculateStats(list.map((r) => r.result.cost_usd)),
    };
  }

  // Numeric delta between NAMED configs.
  const baseline = runSummary[args.baseline_config];
  const candidate = runSummary[args.candidate_config];
  const deltaBlock =
    baseline && candidate
      ? {
          baseline: args.baseline_config,
          candidate: args.candidate_config,
          pass_rate: delta(candidate.pass_rate.mean, baseline.pass_rate.mean),
          time_seconds: delta(candidate.time_seconds.mean, baseline.time_seconds.mean),
          tokens: delta(candidate.tokens.mean, baseline.tokens.mean),
          cost_usd: delta(candidate.cost_usd.mean, baseline.cost_usd.mean),
        }
      : null;
  if (!deltaBlock) {
    notes.push(
      `delta not computed: need both "${args.baseline_config}" and "${args.candidate_config}" configs (found: ${[...byConfig.keys()].join(", ")})`,
    );
  }

  // Gates (code-computed).
  const pressureRuns = runs.filter((r) => r.configuration === args.candidate_config && r.pressure_verdict !== null);
  const pressureAdherence = pressureRuns.length === 0 ? null : pressureRuns.every((r) => r.pressure_verdict === "COMPLIANT");
  const skillLift = baseline && candidate ? candidate.pass_rate.mean > baseline.pass_rate.mean : null;

  const runsPerConfig = Math.max(0, ...[...byConfig.values()].map((l) => l.length));
  const benchmark: Benchmark = {
    metadata: {
      skill_name: args.skill_name,
      skill_path: args.skill_path ?? "",
      timestamp: new Date().toISOString(),
      evals_run: [...new Set(runs.map((r) => r.eval_name))].sort(),
      runs_per_configuration: runsPerConfig,
    },
    runs,
    run_summary: runSummary,
    delta: deltaBlock,
    gates: { pressure_adherence: pressureAdherence, skill_lift: skillLift },
    notes,
  };

  const jsonPath = join(iterDir, "benchmark.json");
  const mdPath = join(iterDir, "benchmark.md");
  writeFileSync(jsonPath, JSON.stringify(benchmark, null, 2));
  writeFileSync(mdPath, generateMarkdown(benchmark));

  const lines: string[] = [
    `aggregated ${runs.length} runs across ${byConfig.size} config(s) and ${benchmark.metadata.evals_run.length} eval(s)`,
  ];
  for (const [config, stats] of Object.entries(runSummary)) {
    lines.push(`${config}: pass_rate mean ${stats.pass_rate.mean} (±${stats.pass_rate.stddev})`);
  }
  if (deltaBlock) {
    lines.push(
      `delta (${deltaBlock.candidate} − ${deltaBlock.baseline}): pass_rate ${deltaBlock.pass_rate >= 0 ? "+" : ""}${deltaBlock.pass_rate}, ` +
        `time ${deltaBlock.time_seconds >= 0 ? "+" : ""}${deltaBlock.time_seconds}s, tokens ${deltaBlock.tokens >= 0 ? "+" : ""}${deltaBlock.tokens}`,
    );
  }
  lines.push(
    `gates: pressure_adherence=${pressureAdherence === null ? "n/a" : pressureAdherence}, skill_lift=${skillLift === null ? "n/a" : skillLift}`,
  );
  for (const note of notes) lines.push(`note: ${note}`);
  lines.push(`wrote ${jsonPath} and ${mdPath}`);
  if (args.response_format === "detailed") {
    lines.push("", "| run | config | pass_rate | pressure | time_s | tokens |", "|---|---|---|---|---|---|");
    for (const r of runs) {
      lines.push(
        `| ${r.eval_id}/run-${r.run_number} | ${r.configuration} | ${r.result.pass_rate} | ${r.pressure_verdict ?? "-"} | ${r.result.time_seconds} | ${r.result.tokens} |`,
      );
    }
  }

  return ok(lines.join("\n"), {
    ...benchmark,
    benchmark_json_path: jsonPath,
    benchmark_md_path: mdPath,
  });
}

// Compile-time contract: handler input is derived from inputShape (defineTool in register.ts re-checks).
void ({ name, title, description, inputShape, handler } satisfies ToolModule<typeof inputShape>);
